### PR TITLE
Removed STB related code

### PR DESCRIPTION
**Pull request by PranavReddyP16**
Placed all stb related code into a new file load_image.cpp based off of issue #2152

### CODE REVIEW

File: src/mlpack/core/data/load_image_impl.hpp [hunk 1]
```diff
@@ -19,8 +19,6 @@
 namespace mlpack {
 namespace data {
 
-#ifdef HAS_STB // Compile this only if stb is present.
```
Comment (himanshupathak21061998) [inline src/mlpack/core/data/load_image_impl.hpp]: Hey @PranavReddyP16 Thanks for contribution. Can you please explain just by removing `#ifdef HAS_STB` is helping here. Because the code present in this file still depends upon `HAS_STB` . In this we are trying to remove depedency upon HAS_STB so that we can use this command https://github.com/mlpack/models/issues/35#issuecomment-572470604 without using -DHAS_STB flag.

Comment (PranavReddyP16) [reply src/mlpack/core/data/load_image_impl.hpp]: Yes, so now the code in this file does not need the HAS_STB macro to be defined for it to work.
Now if we want to load an image, it has nothing to do with the HAS_STB macro.
Was this not the expected output?

Also I have taken all the related code for image loading from this file and placed it in load_image.cpp
Please correct me if I'm wrong.

**Comment by kartikdutt18** [comment]
Hi @PranavReddyP16, your changes fir ISRLU are also shown here, could you remove them.
You can use git revert head to undo commits. My suggestion is that you use master fir rebasing and create branches for different issues. Thanks.

File: src/mlpack/core/data/load_image_impl.hpp (hunk 1 above)
Comment (himanshupathak21061998) [reply src/mlpack/core/data/load_image_impl.hpp]: Hey @PranavReddyP16 If you are able to load an image using your build like using this program.
```
  #include <mlpack/core.hpp>
  #define STB_IMAGE_IMPLEMENTATION
  #include <mlpack/core/data/load_image_impl.hpp>

  using namespace mlpack;
  using namespace mlpack::ann;
  using namespace arma;

  int main()
  {
      std::cout << "Reading data ..." << endl;
      arma::Mat<unsigned char> matrix;
      data::ImageInfo info;
      data::Load("saitama.png", matrix, info, false, true);
      std::cout<<matrix;
  }

```
and compiling this program using  command like this
```
g++ image.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp
```
without using `-DHAS_STB` flag. Then it is good to go otherwise try something else.

Comment (PranavReddyP16) [reply src/mlpack/core/data/load_image_impl.hpp]: Hey @himanshupathak21061998 I'm getting this error when I try to compile with the command you mentioned above: 

Package mlpack was not found in the pkg-config search path.
Perhaps you should add the directory containing `mlpack.pc'
to the PKG_CONFIG_PATH environment variable
Package 'mlpack', required by 'virtual:world', not found
/usr/bin/ld: /tmp/cccxXTii.o: in function `main':
test.cpp:(.text+0x1f7ae): undefined reference to `bool mlpack::data::Load<unsigned char>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, arma::Mat<unsigned char>&, mlpack::data::ImageInfo&, bool, bool)'
collect2: error: ld returned 1 exit status

Any help is appreciated

**Comment by PranavReddyP16** [comment]
@kartikdutt18 Thank you so much for the solution. Will do that

File: src/mlpack/core/data/load_image_impl.hpp (hunk 1 above)
Comment (himanshupathak21061998) [reply src/mlpack/core/data/load_image_impl.hpp]: @PranavReddyP16 Where do you compile it inside build directory or not?

File: src/mlpack/core/data/load_image_impl.hpp [hunk 2]
```diff
@@ -118,30 +116,6 @@ bool Load(const std::vector<std::string>& files,
...
-bool Load(const std::vector<std::string>& files,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal = false,
-          const bool transpose = true)
-{
-  throw std::runtime_error("Load(): HAS_STB is not defined, "
-      "so STB is not available and images cannot be loaded!");
-}
-#endif // HAS_STB.
```
Comment (rcurtin) [inline src/mlpack/core/data/load_image_impl.hpp]: This should be in `load_image.cpp`, since if mlpack itself is compiled without STB support, then these runtime errors should be thrown.

File: src/mlpack/core/data/load_image_impl.hpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/core/data/load_image_impl.hpp]: @himanshupathak21061998 why does `STB_IMAGE_IMPLEMENTATION` need to be `#define`d in your example program?  We should be able to just include `<mlpack/core.hpp>` and `<mlpack/core/data/load_image.hpp>` (if `core.hpp` doesn't already include that) and then your example code should work fine.

@PranavReddyP16 the goal is to make the code work as it currently does, but with any instances of the `HAS_STB` macro (which will still be needed) occuring *only* in the .cpp file.

The program that @himanshupathak21061998 should compile also with the `-DHAS_STB` flag.  When it's compiled with `-DHAS_STB`, it should work and print a matrix; when it's compiled without `-DHAS_STB`, then it should throw an exception when it's run.

File: src/mlpack/methods/ann/activation_functions/CMakeLists.txt [hunk 3]
```diff
@@ -9,6 +9,7 @@ set(SOURCES
   softplus_function.hpp
   swish_function.hpp
   mish_function.hpp
+  isrlu_function.hpp
```
Comment (rcurtin) [inline src/mlpack/methods/ann/activation_functions/CMakeLists.txt]: This change looks unrelated; can you make sure you are developing your work for each PR on a separate branch, so that the changes don't get mixed up?  Thanks. :+1:

File: src/mlpack/core/data/image_info.hpp [hunk 4]
```diff
@@ -28,12 +27,10 @@
 #define STB_IMAGE_WRITE_IMPLEMENTATION
 #include <stb_image_write.h>
 
-#endif
```
Comment (rcurtin) [inline src/mlpack/core/data/image_info.hpp]: Any STB-related code---including the `#include`s---should be in the .cpp file you made and removed in all other files.  So these includes between the `#ifdef` that you removed should also be removed.

File: src/mlpack/core/data/image_info.hpp [hunk 5]
```diff
@@ -44,7 +41,6 @@ namespace data {
 inline bool ImageFormatSupported(const std::string& fileName,
                                  const bool save = false);
 
-#endif
```
Comment (rcurtin) [inline src/mlpack/core/data/image_info.hpp]: If this is only needed when `HAS_STB` is present, you could move it to be a utility function in the .cpp file that you made.

File: src/mlpack/core/data/load_image_impl.hpp (hunk 1 above)
Comment (himanshupathak21061998) [reply src/mlpack/core/data/load_image_impl.hpp]: @rcurtin Sorry, I thought we are trying to change behaviour by refactoring. I removed #define from example.

Comment (rcurtin) [reply src/mlpack/core/data/load_image_impl.hpp]: :+1: it *sort* of changes behavior of the codebase, in that it's no longer necessary to specify `HAS_STB` when compiling.  Also, I realize there is an error I wrote about the program's behavior.  I've edited my comment now.

File: src/mlpack/methods/ann/activation_functions/CMakeLists.txt (hunk 3 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/CMakeLists.txt]: I'm really sorry about this. I'll make seperate branches from now on.

File: src/mlpack/core/data/image_info.hpp (hunk 5 above)
Comment (PranavReddyP16) [reply src/mlpack/core/data/image_info.hpp]: I think my approach to this problem was slightly off. I'll take a look at it once again properly. Thank you for all the advice

**Comment by PranavReddyP16** [comment]
@rcurtin Just to confirm: All the code in the #ifdef HAS_STB has to be removed from current files and placed in load.cpp am I correct?

**Comment by rcurtin** [comment]
@PranavReddyP16 it may need to be further refactored a little bit, but yes, the idea is to ensure that the `HAS_STB` macro is used *only* in the `load.cpp` file.  Thus, because it is in a .cpp file, it will be compiled into `libmlpack.so`, and `HAS_STB` will therefore only matter for the initial compilation of mlpack, not for when any programs are compiled against mlpack.

**Comment by PranavReddyP16** [comment]
@rcurtin Oh I didn't know only cpp files were compiled into libmlpack.so. Also I have deleted this master branch and created a new one to resolve some remaining branch issues I had. I will open another PR regarding this sometime later today

**Comment by rcurtin** [comment]
Sounds good, thanks. For a better understanding of what files are and aren't compiled into the library, I'd suggest spending some time reading about templates and why implementations of templated functions need to be in `.hpp` or _impl.hpp` files. :+1:

**Comment by PranavReddyP16** [comment]
@rcurtin I'm a bit confused here. In load_image_impl.hpp there are a few lines of code which use stb. But in a comment on the issue it was mentioned that such code should be placed in load_image.cpp:

https://github.com/mlpack/mlpack/issues/2152#issuecomment-576818796

 But I was under the impression that templated functions should not be placed in .cpp files. Any advice on how to proceed from here?

**Comment by rcurtin** [comment]
Refactor the code so that the templated function calls a non-templated function that you can put into the .cpp file.

**Comment by PranavReddyP16** [comment]
@rcurtin Can you elaborate on that some more? I don't exactly understand how that's supposed to be done

**Comment by rcurtin** [comment]
I'm not sure how to explain this further.  Wrap the code that calls STB in a separate non-templated function.  Put that code in the .cpp file.  Call that code from the _impl.hpp file.

**Comment by PranavReddyP16** [comment]
Ok yeah I get it. I was just a little confused because I don't think I've seen that anywhere in the codebase before :sweat_smile:

**Comment by jeffin143** [comment]
Please wait for some time before solving this :)
